model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/HotRodMapStorage.java

```diff
@@ -0,0 +1,178 @@
+        this.cloner = cloner;
+    }
+
+    @Override
+    public V create(V value) {
+        K key = keyConvertor.fromStringSafe(value.getId());
+        if (key == null) {
+            key = keyConvertor.yieldNewUniqueKey();
+            value = cloner.from(keyConvertor.keyToString(key), value);
+        }
+
+        remoteCache.putIfAbsent(key, value);
+
+        return value;
+    }
+
+    @Override
+    public V read(String key) {
+        Objects.requireNonNull(key, "Key must be non-null");
+        K k = keyConvertor.fromStringSafe(key);
+        return remoteCache.get(k);
+    }
+
+    @Override
+    public V update(V value) {
+        K key = keyConvertor.fromStringSafe(value.getId());
+        return remoteCache.replace(key, value);
+    }
+
+    @Override
+    public boolean delete(String key) {
+        K k = keyConvertor.fromStringSafe(key);
+        return remoteCache.remove(k) != null;
+    }
+
+    private static String toOrderString(QueryParameters.OrderBy<?> orderBy) {
+        SearchableModelField<?> field = orderBy.getModelField();
+        String modelFieldName = IckleQueryMapModelCriteriaBuilder.getFieldName(field);
+        String orderString = orderBy.getOrder().equals(QueryParameters.Order.ASCENDING) ? "ASC" : "DESC";
+
+        return modelFieldName + " " + orderString;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Stream<V> read(QueryParameters<M> queryParameters) {
+        IckleQueryMapModelCriteriaBuilder<K, V, M> iqmcb = queryParameters.getModelCriteriaBuilder().unwrap(IckleQueryMapModelCriteriaBuilder.class);
+        String queryString = iqmcb.getIckleQuery();
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            queryString += " ORDER BY " + queryParameters.getOrderBy().stream().map(HotRodMapStorage::toOrderString)
+                                            .collect(Collectors.joining(", "));
+        }
+
+        LOG.tracef("Executing read Ickle query: %s", queryString);
+
+        QueryFactory queryFactory = Search.getQueryFactory(remoteCache);
+
+        Query<V> query = paginateQuery(queryFactory.create(queryString), queryParameters.getOffset(),
+                queryParameters.getLimit());
+
+        query.setParameters(iqmcb.getParameters());
+
+        CloseableIterator<V> iterator = query.iterator();
+        return closing(StreamSupport.stream(Spliterators.spliteratorUnknownSize(iterator, 0), false))
+                .onClose(iterator::close);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public long getCount(QueryParameters<M> queryParameters) {
+        IckleQueryMapModelCriteriaBuilder<K, V, M> iqmcb = queryParameters.getModelCriteriaBuilder().unwrap(IckleQueryMapModelCriteriaBuilder.class);
+        String queryString = iqmcb.getIckleQuery();
+
+        if (!queryParameters.getOrderBy().isEmpty()) {
+            queryString += " ORDER BY " + queryParameters.getOrderBy().stream().map(HotRodMapStorage::toOrderString)
```

Comment:
Is ordering required with count operation?

Reply:
This is a good question. We probably don't need it. 

However, there is another thing similar to this for this method. Currently, we are ignoring pagination for `getCount` as [`hitCount`](https://docs.jboss.org/infinispan/11.0/apidocs/org/infinispan/query/dsl/QueryResult.html#hitCount()) method ignores it. We use it for CHM though.

I am not sure whether we need to have pagination for `getCount`, @martin-kanis  @hmlnarik WDYT?

Reply:
For `getCount`, the pagination should be ignored, as the number of the pages is derived from the count

Reply:
Thanks @hmlnarik, in that case, the version of `getCount` that is in this PR should be correct.